include/bio/plain_io/reader.hpp

```diff
@@ -0,0 +1,590 @@
+ */
+template <plain_io::record_kind record_kind_ = plain_io::record_kind::line>
+class plaintext_input_iterator
+{
+private:
+    //!\brief Down-cast pointer to the stream-buffer.
+    bio::detail::stream_buffer_exposer<char> * stream_buf = nullptr;
+
+    //!\brief Place to store lines that overlap buffer boundaries.
+    std::string         overflow_buffer;
+    //!\brief Temporary storage for field delimiter positions.
+    std::vector<size_t> field_end_positions;
+
+    //!\brief The record.
+    plain_io::record record_;
+
+    //!\brief Whether iterator is at end.
+    bool at_end    = false;
+    //!\brief Delimiter between fields.
+    char field_sep = '\t';
+    //!\brief Delimiter between records [not exposed to modification ATM].
+    char rec_sep   = '\n';
+
+    //!\brief Initialisation function.
+    void init(bool read_first)
+    {
+        if (read_first) // read first record
+        {
+            operator++();
+        }
+        else // only ensure streambuffer is readable
+        {
+            assert(stream_buf != nullptr);
+            stream_buf->underflow(); // ensure the stream buffer has content on construction
+            if (stream_buf->gptr() == stream_buf->egptr())
+                at_end = true;
+        }
+    }
+
+public:
+    /*!\name Associated types
+     * \{
+     */
+    using difference_type = ptrdiff_t; //!< Defaults to ptrdiff_t.
+    //!\brief The record type.
+    using value_type =
+      std::conditional_t<record_kind_ == plain_io::record_kind::line, std::string_view, plain_io::record>;
+    //!\brief A reference to the record type.
+    using reference =
+      std::conditional_t<record_kind_ == plain_io::record_kind::line, std::string_view, plain_io::record const &>;
+    using pointer           = value_type const *;      //!< Has no pointer type.
+    using iterator_category = std::input_iterator_tag; //!< Pure input iterator.
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    plaintext_input_iterator() noexcept                                 = default;             //!< Defaulted.
+    plaintext_input_iterator(plaintext_input_iterator const &) noexcept = default;             //!< Defaulted.
+    plaintext_input_iterator(plaintext_input_iterator &&) noexcept      = default;             //!< Defaulted.
+    plaintext_input_iterator & operator=(plaintext_input_iterator const &) noexcept = default; //!< Defaulted.
+    plaintext_input_iterator & operator=(plaintext_input_iterator &&) noexcept = default;      //!< Defaulted.
+    ~plaintext_input_iterator() noexcept                                       = default;      //!< Defaulted.
+
+    //!\brief Construct from a stream buffer.
+    explicit plaintext_input_iterator(std::basic_streambuf<char> & ibuf, bool const read_first = true) :
+      stream_buf{reinterpret_cast<bio::detail::stream_buffer_exposer<char> *>(&ibuf)}
+    {
+        init(read_first);
```

Comment:
```suggestion
        init(read_first_record);
```
It actually threw me off when skipping over it. I didn't immediately thought of the record but `why would you read first, and init later`. Could also mean `read_firat_char` so I would be clearer everywhere.